include/pybind11/detail/internals.h

```diff
@@ -515,7 +535,61 @@ inline local_internals &get_local_internals() {
     // static deinitialization fiasco. In order to avoid it we avoid destruction of the
     // local_internals static. One can read more about the problem and current solution here:
     // https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables
+
+#if defined(PYPY_VERSION) || defined(GRAALVM_PYTHON) || PY_VERSION_HEX < 0x030C0000               \
+    || !defined(PYBIND11_SUBINTERPRETER_SUPPORT)
     static auto *locals = new local_internals();
+#else
+    static thread_local local_internals *locals = nullptr;
+    // This is one per interpreter, we cache it but if the interpreter changed
+    // then we need to invalidate our cache and re-fetch from the state dict
+    static thread_local PyThreadState *tstate_cached = nullptr;
+#    if PY_VERSION_HEX < 0x030D0000
+    PyThreadState *tstate = _PyThreadState_UncheckedGet();
+#    else
+    PyThreadState *tstate = PyThreadState_GetUnchecked();
+#    endif
+    if (!tstate) {
+        pybind11_fail(
+            "pybind11::detail::get_local_internals() called without a current python thread");
+    }
+    if (tstate != tstate_cached) {
+        // we create a unique value at first run which is based on a pointer to
+        // a (non-thread_local) static value in this function, then multiple
+        // loaded modules using this code will still each have a unique key.
+        static const std::string this_module_idstr
+            = PYBIND11_MODULE_LOCAL_ID
+              + std::to_string(reinterpret_cast<uintptr_t>(&this_module_idstr));
+
+        // Ensure that the GIL is held since we will need to make Python calls.
+        // Cannot use py::gil_scoped_acquire here since that constructor calls get_internals.
+        struct gil_scoped_acquire_local {
+            gil_scoped_acquire_local() : state(PyGILState_Ensure()) {}
+            gil_scoped_acquire_local(const gil_scoped_acquire_local &) = delete;
+            gil_scoped_acquire_local &operator=(const gil_scoped_acquire_local &) = delete;
+            ~gil_scoped_acquire_local() { PyGILState_Release(state); }
+            const PyGILState_STATE state;
+        } gil;
```

Comment:
That's a lot of copy-paste. We should clean this up, maybe with a new pybind11/detail/gil_simple.h, that we reuse here, and from pybind11/gil.h. Could you please help by sending a PR for that? I'll review that one quickly.

Reply:
I moved this up to avoid the copy and paste.